### PR TITLE
feat(Content): Mass import of recipes from cookipedia

### PR DESCRIPTION
**Pull request by dawehner**
Task: #

* [x] Ready for review
* [x] Ready for merge

- Uses cookipedia.co.uk as source for more recipes
- Provides a script to fetch their "favourite recipes"
- As mediawiki pages are not API first we need to scrape some of the data (like the difficulty) from the HTML itself, meh
- Adapts the authors migration to not require a custom authors.csv file, but rather we now take the author names from the recipes.csv file, much like the categories and tags

Todo:

- [x] Figure out how to reference remote images. We probably don't want to download 200 images, everytime we install the distribution?
- [ ] (Figure out why we the script does not always fetches the difficulty)

### Notes

... Notes-optional

### Test Instructions

Test-instructions

### QA

... Screenshots/Screencasts-optional

### CODE REVIEW

File: modules/recipes_magazin/migrations/recipes_magazin__authors.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ id: recipes_magazin__authors
 status: true
 source:
   plugin: recipe_magazin__terms
-  path: default_content/authors.csv
+  path: default_content/recipes.csv
```
Comment (e0ipso) [inline modules/recipes_magazin/migrations/recipes_magazin__authors.yml:5]: Yeah

**Comment by e0ipso** [comment]
This looks good to me. We still have to figure out the images thing for the data model to work together. Luckily all images are remote in the decoupled world, so consumers won't notice it.

**Comment by e0ipso** [comment]
Another good alternative can be to serve those recipes with a default image and have an optional zip with 200 jpgs ready to insert. That seems like a lot of work though.

**Comment by e0ipso** [comment]
Another solution is to download 20 recipes with the 20 images.

File: modules/recipes_magazin/src/Plugin/migrate/source/RecipeExternalImages.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace Drupal\recipes_magazin\Plugin\migrate\source;
+
+use Drupal\Component\Utility\UrlHelper;
+
+/**
+ * Extends the CSV migration to make it possible to import terms.
```
Comment (e0ipso) [inline modules/recipes_magazin/src/Plugin/migrate/source/RecipeExternalImages.php:8]: _[…] it possible to import external images_.

Comment (dawehner) [reply modules/recipes_magazin/src/Plugin/migrate/source/RecipeExternalImages.php:8]: ha, fair point

**Comment by dawehner** [comment]
Sadly currently I have a hard time figuring out how to deal with images which are remote.

Here is a backtrace of the problem we see:
```

filesize(): stat failed for https://www.cookipedia.co.uk/wiki/images/7/73/Pork_fillet%2C_apple_and_sweet_potato_bake.jpg in  [warning]
Drupal\Core\Image\Image-&gt;__construct() (line 54 of core/lib/Drupal/Core/Image/Image.php).
Drupal\Core\Image\Image-&gt;__construct(Object,
&#039;https://www.cookipedia.co.uk/wiki/images/7/73/Pork_fillet%2C_apple_and_sweet_potato_bake.jpg&#039;) (Line: 85)
Drupal\Core\Image\ImageFactory-&gt;get(&#039;https://www.cookipedia.co.uk/wiki/images/7/73/Pork_fillet%2C_apple_and_sweet_potato_bake.jpg&#039;)
(Line: 317)
Drupal\image\Plugin\Field\FieldType\ImageItem-&gt;preSave() (Line: 244)
Drupal\Core\Field\FieldItemList-&gt;delegateMethod(&#039;preSave&#039;) (Line: 202)
Drupal\Core\Field\FieldItemList-&gt;preSave() (Line: 479)
Drupal\Core\Entity\ContentEntityStorageBase-&gt;invokeFieldMethod(&#039;preSave&#039;, Object) (Line: 429)
Drupal\Core\Entity\ContentEntityStorageBase-&gt;invokeHook(&#039;presave&#039;, Object) (Line: 435)
Drupal\Core\Entity\EntityStorageBase-&gt;doPreSave(Object) (Line: 298)
Drupal\Core\Entity\ContentEntityStorageBase-&gt;doPreSave(Object) (Line: 389)
Drupal\Core\Entity\EntityStorageBase-&gt;save(Object) (Line: 768)
Drupal\Core\Entity\Sql\SqlContentEntityStorage-&gt;save(Object) (Line: 364)
Drupal\Core\Entity\Entity-&gt;save() (Line: 109)
Drupal\migrate\Plugin\migrate\destination\EntityContentBase-&gt;save(Object, Array) (Line: 90)
Drupal\migrate\Plugin\migrate\destination\EntityContentBase-&gt;import(Object, Array) (Line: 224)
Drupal\migrate\MigrateExecutable-&gt;import()
call_user_func_array(Array, Array) (Line: 67)
Drupal\recipes_magazin\MigrationRunner-&gt;Drupal\recipes_magazin\{closure}(&#039;recipes_magazin__media_images_url&#039;, 3)
array_walk(Array, Object) (Line: 68)
Drupal\recipes_magazin\MigrationRunner-&gt;execute(Array, &#039;import&#039;) (Line: 39)
Drupal\recipes_magazin\MigrationRunner-&gt;run(Array) (Line: 13)
recipes_magazin_install()
call_user_func_array(&#039;recipes_magazin_install&#039;, Array) (Line: 391)
Drupal\Core\Extension\ModuleHandler-&gt;invoke(&#039;recipes_magazin&#039;, &#039;install&#039;) (Line: 287)
Drupal\Core\Extension\ModuleInstaller-&gt;install(Array, 1) (Line: 83)
Drupal\Core\ProxyClass\Extension\ModuleInstaller-&gt;install(Array) (Line: 40)
contenta_jsonapi_install_configure_form_submit(Array, Object)
call_user_func_array(&#039;contenta_jsonapi_install_configure_form_submit&#039;, Array) (Line: 111)
Drupal\Core\Form\FormSubmitter-&gt;executeSubmitHandlers(Array, Object) (Line: 51)
Drupal\Core\Form\FormSubmitter-&gt;doSubmitForm(Array, Object) (Line: 585)
Drupal\Core\Form\FormBuilder-&gt;processForm(&#039;install_configure_form&#039;, Array, Object) (Line: 481)
Drupal\Core\Form\FormBuilder-&gt;submitForm(&#039;Drupal\Core\Installer\Form\SiteConfigureForm&#039;, Object) (Line: 938)
install_get_form(&#039;Drupal\Core\Installer\Form\SiteConfigureForm&#039;, Array) (Line: 594)
install_run_task(Array, Array) (Line: 550)
install_run_tasks(Array) (Line: 121)
install_drupal(Object, Array) (Line: 726)
drush_call_user_func_array(&#039;install_drupal&#039;, Array) (Line: 711)
drush_op(&#039;install_drupal&#039;, Object, Array) (Line: 82)
drush_core_site_install_version(&#039;contenta_jsonapi&#039;, Array) (Line: 255)
drush_core_site_install(&#039;contenta_jsonapi&#039;) (Line: 422)
_drush_invoke_hooks(Array, Array) (Line: 231)
drush_command(&#039;contenta_jsonapi&#039;) (Line: 199)
drush_dispatch(Array) (Line: 67)
Drush\Boot\BaseBoot-&gt;bootstrap_and_dispatch() (Line: 66)
drush_main() (Line: 12)

```

 I found a module which is doing that: https://www.drupal.org/project/remote_image but this means the data structure changes, aka. we no longer use file entities, but rather just media entities with a URL to the image directly. Obviously, this is not a solution ...

**Comment by dawehner** [comment]
I guess we need to take the pill and download the images somehow. I suggest to have stub versions of the images being created, and then maybe an additional command to actually download the files.

**Comment by e0ipso** [comment]
Sadly there was an unforeseen problem that makes this route unusable for images. But the rest of the PR is good to go.

**Comment by dawehner** [comment]
We solved the unforseen problem by using remote_stream_wrapper (+ a small little patch to fix its dependencies).

**Comment by dawehner** [comment]
❤️

**Pull request merged**